firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift

```diff
@@ -1725,6 +1734,68 @@ class BrowserViewController: UIViewController,
         } else if state.showQRcodeReader {
             navigationHandler?.showQRCode(delegate: self)
         }
+
+        handleNavigationActions(for: state)
+    }
+
+    private func dispatchBackForwardToolbarAction(_ isEnabled: Bool?, _ windowUUID: UUID, _ actionType: ToolbarActionType) {
+        let action = ToolbarAction(isEnabled: isEnabled, windowUUID: windowUUID, actionType: actionType)
+
+        switch actionType {
+        case .backButtonStatus,
+             .forwardButtonStatus:
+            store.dispatch(action)
+        default: break
+        }
+    }
+
+    private func getNavigationAction(for state: BrowserViewControllerState,
+                                     actionType: ToolbarState.ActionState.ActionType) -> Bool {
+        let navigationToolbarAction = actionType == .back ?
+            state.toolbarState.navigationToolbar.actions.first(where: { $0.actionType == actionType })?.navigateBack :
+            state.toolbarState.navigationToolbar.actions.first(where: { $0.actionType == actionType })?.navigateForward
+
+        let addressToolbarAction = actionType == .back ?
+            state.toolbarState
+                 .addressToolbar
+                 .navigationActions
+                 .first(where: { $0.actionType == actionType })?.navigateBack :
+
+            state.toolbarState
+                 .addressToolbar
+                 .navigationActions
+                 .first(where: { $0.actionType == actionType })?.navigateForward
+
+        return navigationToolbarAction ?? addressToolbarAction ?? false
+    }
+
+    private func handleNavigationActions(for state: BrowserViewControllerState) {
+        let navigateBack = getNavigationAction(for: state, actionType: .back)
+        if navigateBack {
+            dismissUrlBar()
+            updateZoomPageBarVisibility(visible: false)
+            tabManager.selectedTab?.goBack()
```

Comment:
curious - why do these actions not occur due to the change in state and occurs before the dispatch?

Reply:
The actions actually take place after the state has been changed. This is handled in the `ToolbarMiddleware` class, specifically within the `handleToolbarButtonTapActions` method. The 'after dispatch' is utilized to alter the navigation state to false. This is done to prevent an infinite cycle, a strategy similar to what is implemented in the `didTapOnHome()` function.